elf/arch-loongarch.cc

```diff
@@ -755,6 +762,31 @@ void shrink_section(Context<E> &ctx, InputSection<E> &isec, bool use_rvc) {
           sign_extend(val, 11) == val)
         delta += 4;
       break;
+    case R_LARCH_PCALA_HI20:
+      if (i + 4 > rels.size())
+        continue;
+
+      ul32 pcala = *(ul32 *)(isec.contents.data() + rels[i].r_offset);
+      ul32 addi = *(ul32 *)(isec.contents.data() + rels[i+2].r_offset);
+
+      i64 rd = pcala & 0x1f;
+      const ul32 addi_d = 0x02c00000;
+      const u64 loc = isec.get_addr() + r.r_offset - delta;
+
+      u64 symval = sym.get_addr(ctx) + r.r_addend;
+      /* Is pcalau12i + addi.d insns?  */
+      if (rels[i+2].r_type != R_LARCH_PCALA_LO12
+          || rels[i+3].r_type != R_LARCH_RELAX
+          || (addi & addi_d) != addi_d
```

Comment:
By the way, this line is wrong as a test if the instruction is `addi` right? I took a look at the LoongArch instruction format, and it says that the opcode always starts at the most significant bit. So it looks like it should be `addi & 0xffc0'0000) != addi`.

Reply:
> By the way, this line is wrong as a test if the instruction is `addi` right? I took a look at the LoongArch instruction format, and it says that the opcode always starts at the most significant bit. So it looks like it should be `(addi & 0xffc0'0000) != addi`.

You are right, this is a mistake.